Detect if access to localStorage is forbidden by the user's browser

thanatos opened the pull request:
If the user's cookie/persistent storage setting forbid access to `localStorage`, catch the exception and abort the access.

Currently, attempting to use the expand/contract links at the top of the page for structs/consts/etc. fails due to an unhandled error while accessing `localStorage`, if such access is forbidden, as the exception from the failed access propagates all the way out, interrupting the expand/contract. Instead, I would like to degrade gracefully; the access won't happen (the collapse/expand state won't get persisted) but the actual expanding/contracting of the item will go on to succeed.

Fixes #55079

rust-highfive commented:
Some changes occurred in HTML/CSS.

cc @GuillaumeGomez

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @GuillaumeGomez (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

thanatos commented:
Note: I'm not sure how to actually test this change, so I haven't done that yet.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/441440186) ([raw log](https://api.travis-ci.org/v3/job/441440186/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:11:11] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:31: tab character
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:32: tab character
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:33: tab character
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:34: tab character
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:35: tab character
[00:11:12] tidy error: /checkout/src/librustdoc/html/static/storage.js:36: tab character
[00:11:13] some tidy checks failed
[00:11:13] 
[00:11:13] 
[00:11:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:11:13] 
[00:11:13] 
[00:11:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:11:13] Build completed unsuccessfully in 0:00:46
[00:11:13] Build completed unsuccessfully in 0:00:46
[00:11:13] make: *** [tidy] Error 1
[00:11:13] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:04bca9c8
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:0213adc0:start=1539564062655847355,finish=1539564062663845416,duration=7998061
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:07826e80
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0dbff480
travis_time:start:0dbff480
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:31a79876
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

GuillaumeGomez commented:
Thanks for the PR! Please update the "getCurrentValue" function (just below the one you updated) too.

thanatos commented:
Given the way `getCurrentValue` was written, it was much easier to update that function if the check itself was a separate function, so I've extracted the code to a separate function. (Which, honestly, I think was the right thing overall, and makes the update function nicer too.)

LMK if there's anything else needed; also, feel free to squash/not squash however the rust-lang folks like to do.

I ran `cargo doc` in a project of mine and replaced most of the `storage.js` that `cargo` emitted w/ the file as it is in my revision. (The exception being that the one `cargo` emits contains,
```javascript
var resourcesSuffix="";
```
just prior to the copyright banner; I have no idea where that comes from, so I left it for my test, only replacing the stuff below the copyright banner. If there's some more rigorous test I could perform, LMK, and I'll do that too.)

GuillaumeGomez commented on src/librustdoc/html/static/storage.js:
```diff
@@ -27,20 +27,38 @@ function onEach(arr, func) {
     return null;
 }
 
+function usableLocalStorage() {
```
Can move this function above the two others please? It's just me being careful but better be safe than sorry. :)

GuillaumeGomez replied:
Try this code then:

```
foo();

function foo() {
    console.log('hello');
}
```

GuillaumeGomez commented:
Thanks!

@bors: r+ rollup

bors commented:
:pushpin: Commit cbe98ec803c023deb6d200b6761acd7c7d73254c has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez cbe98ec803c023deb6d200b6761acd7c7d73254c -->

kornelski commented:
I think it'd be safer and more future-proof to wrap `getCurrentValue` and `updateLocalStorage` body `try`/`catch`.

For example, if browsers implemented read-only localStorage current check would pass, but write could still throw.

GuillaumeGomez commented:
The write can only be performed on the setting page, therefore it's not really an issue if it doesn't work. The setting will just not get updated.
